### PR TITLE
Add colorMapper util

### PR DESCRIPTION
**Pull request by micha-f**
A simple way to assign colors to strings deterministically without ending up with the same colors for similar names / names starting with the same initials. To be used with avatars and name tags.

### CODE REVIEW

**Comment by TejasQ** [comment]
How does it work?

**Comment by micha-f** [comment]
It uses a hashing function (FNV 1a - https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function) to turn a UTF8 string into a positive int in the 32-bit space (0 to 2147483647*2). It then assigns this number to one of `n` colors using `mod n`. Small changes in the input string will result in a different hash (int) and thus (most likely) in a different color.

**Comment by TejasQ** [comment]
And there's no simpler/better solution to do this?

**Comment by micha-f** [comment]
If you want to 
- assign strings deterministically to the same color
- assign a different color even for very similar strings (those that only differ in one character)
then you need to somehow turn a string into a hash with a uniform distribution and map that to a color.

I don't think there is a simpler / better solution - otherwise I would have proposed it 😉 .

**Comment by peterszerzo** [comment]
@micha-f I like the idea of doing this properly. If you don't mind, I'll fix up the branch so tests are passing myself, and see how we integrate it to the rest of the codebase.

File: packages/components/src/utils/color.ts [hunk 1]
```diff
@@ -1,2 +1,84 @@
...
+/*
+ * Hashing algorithm used to pair strings with colors (see `colorMapper` above), chosen
+ * as a compromise between speed and low risk of having identical colors assigned for two
+ * strings that are nearly identical.
+ *
+ * Copied from https://github.com/tjwebb/fnv-plus (MIT licensed), author Travis Webb <me@traviswebb.com>
+ * This FNV-1a hash algorithm, often simply called "fnv", disperses hashes throughout
+ * the 32-bit hash space with very good dispersion and is very fast.
+ */
+function hash32FNV1aUTF(str: string) {
```
Comment (TejasQ) [inline packages/components/src/utils/color.ts]: Maybe we can `export const hash = ` this.

Comment (micha-f) [reply packages/components/src/utils/color.ts]: Do we need this in other places? If so - wouldn't it make more sense to move it out of the **color** utils? If we don't have another use case, does it make sense to keep it private as an implementation detail for now?

Comment (TejasQ) [reply packages/components/src/utils/color.ts]: Right right. My mistake. I thought this was in a different file than `color.ts`. This looks good.

File: packages/components/src/utils/__tests__/color.test.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import * as React from "react"
```
Comment (fabien0102) [inline packages/components/src/utils/__tests__/color.test.tsx]: Why do you need react?

**Comment by TejasQ** [comment]
Sweet. A rebase and let's merge.

File: packages/components/src/utils/color.ts [hunk 3]
```diff
@@ -1,2 +1,84 @@
...
+  let v0 = 0x9dc5
+  let t1 = 0
+  let v1 = 0x811c
+
+  for (i = 0; i < l; i = i + 1) {
+    c = str.charCodeAt(i)
+    if (c < 128) {
+      v0 ^= c
+    } else if (c < 2048) {
+      v0 ^= (c >> 6) | 192
```
Comment (fabien0102) [inline packages/components/src/utils/color.ts]: So nice 😍 I like binary operators in js 💯

File: packages/components/src/utils/__tests__/color.test.tsx (hunk 2 above)
Comment (micha-f) [reply packages/components/src/utils/__tests__/color.test.tsx]: Can be removed

**Comment by fabien0102** [comment]
It look very nice but where is it used?

**Comment by TejasQ** [comment]
LGTM

**Comment by fabien0102** [comment]
@peterszerzo can you rebase to fix the netlify build?

File: packages/components/src/NameTag/NameTag.tsx [hunk 4]
```diff
@@ -15,7 +17,10 @@ export interface Props {
    * Indicates that this component is right of other content, and adds an appropriate left margin.
    */
   right?: boolean
-  children?: React.ReactNode
+  /**
+   * Children to this component are expected to be a plain string
+   */
+  children?: string
```
Comment (fabien0102) [inline packages/components/src/NameTag/NameTag.tsx]: Even if is a plain string, I prefer `React.ReactNode` to be able to use `react-intl` easily.

```jsx
import { FormattedMessage } from "react-intl"

export default = () =>
   <NameTag><FormattedMessage id="myTag" defaultMessage="my tag" /></NameTag>
```

versus

```jsx
import { injectIntl, InjectedIntlProps } from "react-intl"

const MyTag: React.SFC<InjectedIntlProps> = ({intl}) =>
  <NameTag>{intl.formatMessage({id: "myTag", defaultMessage: "my tag"})}</NameTag>

export default injectIntl(MyTag)
```

Comment (fabien0102) [reply packages/components/src/NameTag/NameTag.tsx]: For a NameTag is not really important, but it's more as a general rule :wink: (I have play a bit with `react-intl` and `operational`, and it's a bit annoying to have to inject the `intl` everytime

Comment (peterszerzo) [reply packages/components/src/NameTag/NameTag.tsx]: I think it is important to communicate that only strings are allowed here.

It is likely that several components will have to get refactored if we wind up supporting internationalization in this codebase, but luckily those will be pretty low-risk.

**Pull request merged**